repair.py

```diff
@@ -0,0 +1,162 @@
+import asyncio
+import logging
+import sys
+import tempfile
+
+import createrepo_c as cr
+import shutil
+from django.core.files.storage import default_storage as storage
+
+from pulpcore.plugin.models import (
+    ContentArtifact,
+    RepositoryVersion,
+)
+
+from pulp_rpm.app.models import (
+    Package,
+    RpmPublication,
+    RpmRemote,
+    RpmRepository,
+    RpmDistribution,
+)
+
+log = logging.getLogger(__name__)
+loop = asyncio.get_event_loop()
+
+broken_criteria = {
+    "files": [],
+    "changelogs": [],
+}
+broken_packages = Package.objects.filter(**broken_criteria)
+
+print("""
+##########################################
+#  Repair tool for Pulp RPM issue #9107  #
+#    https://pulp.plan.io/issues/9107    #
+##########################################
+""")
+
+print(
+"""Analysis
+========
+"""
+)
+
+total_packages = Package.objects.all().count()
+total_broken_packages = broken_packages.count()
+print("Total broken packages: {}/{}".format(total_broken_packages, total_packages))
+
+broken_content_artifacts = ContentArtifact.objects.filter(content__in=broken_packages)
+assert broken_content_artifacts.count() == total_broken_packages
+
+broken_content_artifacts_with_files = broken_content_artifacts.filter(artifact__isnull=False)
+broken_content_artifacts_without_files = broken_content_artifacts.filter(artifact__isnull=True)
+
+total_broken_packages_with_artifacts = broken_content_artifacts_with_files.count()
+print("Broken packages with local RPMs available: {}/{}".format(total_broken_packages_with_artifacts, total_broken_packages))
+
+total_broken_packages_download_needed = total_broken_packages - total_broken_packages_with_artifacts
+print("Broken packages without local RPMs available: {}/{}".format(total_broken_packages_download_needed, total_broken_packages))
+
+print(
+"""
+Repair
+======
+"""
+)
+
+total_packages_repaired = 0
+
+
+def update_total():
+    global total_packages_repaired
+    global total_broken_packages
+    sys.stdout.write("\rPackages repaired: {}/{}".format(total_packages_repaired, total_broken_packages))
+    sys.stdout.flush()
+
+# Repair on-disk Packages
+
+packages_to_update = []
+for ca in broken_content_artifacts_with_files.select_related("artifact", "content").iterator():
+    with ca.artifact.file as fp:
```

Comment:
Does this need to go through the storage abstraction or will it work as-is?

Reply:
I think so, it seems pretty similar to the code we had to do it:
https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/shared_utils.py#L7-L29